src/Riok.Mapperly/Descriptors/MappingBuilders/SpanMappingBuilder.cs

```diff
@@ -0,0 +1,270 @@
+        var typedArray = ctx.Types.GetArrayType(enumeratedType);
+        if (ctx.FindOrBuildMapping(typedArray.NonNullable(), ctx.Target) is not { } arrayMapping)
+            return null;
+
+        return new SourceObjectMethodDelegateMapping(ctx.Source, ctx.Target, ToArrayMethodName, arrayMapping);
+    }
+
+    private static bool IsTargetImmutableCollection(MappingBuilderContext ctx)
+    {
+        if (SymbolEqualityComparer.Default.Equals(ctx.Target.OriginalDefinition, ctx.Types.Get(typeof(ImmutableArray<>))))
+            return true;
+
+        if (
+            SymbolEqualityComparer.Default.Equals(ctx.Target.OriginalDefinition, ctx.Types.Get(typeof(ImmutableList<>)))
+            || SymbolEqualityComparer.Default.Equals(ctx.Target.OriginalDefinition, ctx.Types.Get(typeof(IImmutableList<>)))
+        )
+            return true;
+
+        if (
+            SymbolEqualityComparer.Default.Equals(ctx.Target.OriginalDefinition, ctx.Types.Get(typeof(ImmutableHashSet<>)))
+            || SymbolEqualityComparer.Default.Equals(ctx.Target.OriginalDefinition, ctx.Types.Get(typeof(IImmutableSet<>)))
+        )
+            return true;
+
+        if (
+            SymbolEqualityComparer.Default.Equals(ctx.Target.OriginalDefinition, ctx.Types.Get(typeof(ImmutableQueue<>)))
+            || SymbolEqualityComparer.Default.Equals(ctx.Target.OriginalDefinition, ctx.Types.Get(typeof(IImmutableQueue<>)))
+        )
+            return true;
+
+        if (
+            SymbolEqualityComparer.Default.Equals(ctx.Target.OriginalDefinition, ctx.Types.Get(typeof(ImmutableStack<>)))
+            || SymbolEqualityComparer.Default.Equals(ctx.Target.OriginalDefinition, ctx.Types.Get(typeof(IImmutableStack<>)))
+        )
+            return true;
+
+        if (SymbolEqualityComparer.Default.Equals(ctx.Target.OriginalDefinition, ctx.Types.Get(typeof(ImmutableSortedSet<>))))
+            return true;
+
+        return false;
+    }
+
+    private static (CollectionType TypeOfCollection, ITypeSymbol EnumeratedType)? GetCollectionInfo(
+        ITypeSymbol symbol,
+        MappingBuilderContext ctx
+    )
+    {
+        var enumeratedType = ctx.Types.GetEnumeratedType(symbol);
+        if (enumeratedType != null)
+        {
+            if (symbol.IsArrayType())
+                return (CollectionType.Array, enumeratedType);
+
+            return (CollectionType.Enumerable, enumeratedType);
+        }
+
+        if (symbol is not INamedTypeSymbol namedTypeSymbol || namedTypeSymbol.TypeArguments.Length != 1)
+            return null;
+
+        var typeSymbol = namedTypeSymbol.TypeArguments[0];
+        if (SymbolEqualityComparer.Default.Equals(symbol.OriginalDefinition, ctx.Types.Get(typeof(Span<>))))
+            return (CollectionType.Span, typeSymbol);
+
+        if (SymbolEqualityComparer.Default.Equals(symbol.OriginalDefinition, ctx.Types.Get(typeof(ReadOnlySpan<>))))
+            return (CollectionType.ReadOnlySpan, typeSymbol);
```

Comment:
I think I'd move the Span/ReadOnlySpan before the array / enumerable checks because it seems to be more intuitive to me. It is not entirely clear that `GetEnumeratedType` just works on `IEnumerable<>` implementations but not on types implementing `GetEnumerator` which would be implemented by span. Therefore moving these checks before the array/enumerable check IMO would make it a lot easier to understand.

Reply:
While addressing https://github.com/riok/mapperly/issues/498 and https://github.com/riok/mapperly/issues/499 I tried sth similar here: https://github.com/riok/mapperly/pull/501.

Reply:
Nice, looks super helpful. Should I add the various `Span` & `Memory` types to it? Not sure if there is enough overlap to justify it.

Reply:
I think adding `Span`, `Memory`, ... makes sense as the checks etc. are the same and these mappings are usually either from and to a span/memory or include at least one collection.